Fix System.Diagnostics.TraceSource tests to stop showing debug assert windows in desktop

safern opened the pull request:
Some of the tests where showing Assert failures windows because in desktop trace listeners `AssertUiEnabled` is set to true by default. 

I also fixed some tests small compat issues with desktop so that they pass. 

cc: @tarekgh @danmosemsft @weshaggard

stephentoub commented on src/System.Diagnostics.TraceSource/tests/DefaultTraceListenerClassTests.cs:
```diff
@@ -27,6 +28,11 @@ public override void Write(string message)
+            public override void WriteLine(string message)
+            {
+                _writer.WriteLine(message);
+            }
```
Why is this override needed?

safern replied:
Because `DefaultTraceListener.Fail(string)` in desktop calls `WriteLine` instead of `Write` which we call in Core

stephentoub replied:
What prevents us from changing core to match desktop?
cc: @danmosemsft

safern replied:
So we have the same behavior but the difference is that in desktop we have a private `Write` method which does the same as the public `Write` method in Core. So they both end up calling `Write` from `DefaultTraceListener.Fail(string)` but in desktop we call the private method which can't be overridden.

safern replied:
The public method `Write` in desktop, just calls the private `Write` API. We deleted that step in Core and just do what the private API did inside the public one. Was I clear?

tarekgh replied:
I talked offline with @safern and we think it is better to have the same desktop behavior. so @safern is going to fix the library and then fix the test to test the expected behavior.

safern replied:
Yes I will fix that and update the PR.

stephentoub replied:
Great, thanks

stephentoub commented on src/System.Diagnostics.TraceSource/tests/TestTraceListener.cs:
```diff
@@ -119,5 +119,10 @@ public override void WriteLine(string message)
+
+        public override void Close()
+        {
+            Dispose(); // In desktop Trace.Close() calls listener.Close() on every listener and in Core we call listener.Dispose() so the tests would fail in desktop without this override. 
```
Should we fix that in the implementation rather than mask it in the test?

safern replied:
We could do that, would you prefer me to open an issue to fix that in desktop and skip the test in desktop?

stephentoub replied:
I meant why not fix the implementation in core to match the implementation in desktop.

safern replied:
In Core when we call `Trace.Close()` we call `listener.Dispose()` and in desktop `listener.Close()`. Probably we rather dispose the objects? Because `TraceListener.Close()` is virtual, so maybe we wanted to do that so that if classes that inherit `TraceListener` don't override `Close` correctly we've taken care of disposing the object.

stephentoub replied:
When in doubt, match what desktop does. :smile:

safern replied:
Sounds something reasonable. Ok, so I'll update it to match what desktop does and update the tests to actually count the number of calls to Close as well. Thanks.

stephentoub commented on src/System.Diagnostics.TraceSource/tests/TraceClassTests.cs:
```diff
@@ -11,6 +11,7 @@ namespace System.Diagnostics.TraceSourceTests
     public class TraceClassTests : IDisposable
     {
         private const string TestRunnerAssemblyName = "xunit.console.netcore";
+        private const string TestRunnerAssemblyNameDesktop = "xunit.console.exe";
```
How come there's a ".exe" on the desktop one but not on the other?

stephentoub replied:
Actually, why do we need this at all?

tarekgh commented on src/System.Diagnostics.TraceSource/tests/TraceSourceClassTests.cs:
```diff
@@ -131,11 +131,20 @@ public void NullSourceName()
+        public void EmptySourceName_Desktop()
+        {
+            Assert.Throws<ArgumentException>(() => new TraceSource(string.Empty));
+            Assert.Throws<ArgumentException>(() => new TraceSource(string.Empty, SourceLevels.All));
```
instead of splitting it into 2 methods, you can change the test code to capture the exception and then validate the exception according to the property 

        internal static bool IsFullFramework => RuntimeInformation.FrameworkDescription.StartsWith(".NET Framework", StringComparison.OrdinalIgnoreCase);

safern replied:
Sounds good.

tarekgh commented on src/System.Diagnostics.TraceSource/tests/TraceClassTests.cs:
```diff
@@ -341,9 +351,42 @@ public void TraceTest01()
         }
 
         [Fact]
```
instead of duplicating the method you can use 
        internal static bool IsFullFramework => RuntimeInformation.FrameworkDescription.StartsWith(".NET Framework", StringComparison.OrdinalIgnoreCase);

to know we are running on desktop and use the desktop runner name. still the idea of dynamically requesting the exe name is valid idea too

weshaggard replied:
FYI - That check for IsFullFramework is not the best way to determine it. If we really need to know that we are on desktop we should be looking at the target framework from AppContext.TargetFrameworkName which has a much better defined contract for this and can even be parsed by TargetFramework type. FrameworkDescription is very loosely defined and is subject to change.

stephentoub replied:
> If we really need to know that we are on desktop we should be looking at the target framework from AppContext.TargetFrameworkName which has a much better defined contract for this and can even be parsed by TargetFramework type. FrameworkDescription is very loosely defined and is subject to change.

Why do we have FrameworkDescription then?  Or why doesn't FrameworkDescription just return AppContext.TargetFrameworkName?

safern replied:
Thanks for the heads up!

weshaggard replied:
FrameworkDescription is only for diagnostic purposes it is not intended to be used for runtime decisions.

stephentoub replied:
Does FrameworkDescription provide more details than TargetFrameworkName?

stephentoub replied:
I guess the difference is that FrameworkDescription includes a version number and TargetFrameworkName doesn't?

safern replied:
Just FYI: we can't use AppContext.TargetFrameworkName yet as it is not available in .NET 46 so we would break the desktop tests in runtime.

weshaggard replied:
TargetFrameworkName follows a well formed structure that can be parsed by things like https://github.com/dotnet/corefx/blob/master/src/System.Runtime.Extensions/src/System/Runtime/Versioning/FrameworkName.cs#L10 and can be specified by https://github.com/dotnet/coreclr/blob/master/src/mscorlib/src/System/Runtime/Versioning/TargetFrameworkAttribute.cs. 

Where as FrameworkDescription doesn't follow any common schema or structure and changes from framework to framework and might be changed at any time. It is only useful for logs and diagnostics purposes.

weshaggard replied:
I just chatted with @safern offline and AppContext.TargetFrameworkName is only available in .NET Core and .NET Framework 4.6.3/4.7 so it isn't useful for our purposes just yet. We should consider adding something similar with the proposal to update RuntimeInformation at https://github.com/dotnet/corefx/issues/16629

tarekgh replied:
right, I looked at  that before and learned from @AlexGhiondea AppContext check will not work pre-4.7. 

I had some other workaround by checking where corelib location but this still a hack. RuntimeInformation would be the best here.

safern commented:
Addressed feedback and updated the PR

tarekgh commented:
LGTM. thanks @safern

stephentoub commented on src/System.Diagnostics.TraceSource/src/System/Diagnostics/DefaultTraceListener.cs:
```diff
@@ -115,7 +115,34 @@ private void WriteAssert(string stackTrace, string message, string detailMessage
         public override void Write(string message)
         {
-            if (NeedIndent) WriteIndent();
+            Write(message, true);
```
Nit: for cases where Boolean arguments are passed, or other non-descript values, it's helpful where possible to use parameter names at the call site, e.g.
```C#
Write(message, useLogFile: true);
```

safern replied:
Ok thanks!

stephentoub commented on src/System.Diagnostics.TraceSource/src/System/Diagnostics/DefaultTraceListener.cs:
```diff
@@ -115,7 +115,34 @@ private void WriteAssert(string stackTrace, string message, string detailMessage
+                WriteIndent();
+
+            // I do the concat here to make sure it goes as one call to the output.
+            // we would save a stringbuilder operation by calling Write twice.
```
Nit: I realize you just copied this comment from the original, it could be cleaned up slightly, though, e.g.
```C#
// The concat is done here to enable a single call to Write
```
I have a small pet peeve against of "I" in shared codebases.

safern replied:
Sounds good.

stephentoub commented on src/System.Diagnostics.TraceSource/src/System/Diagnostics/DefaultTraceListener.cs:
```diff
@@ -132,20 +159,28 @@ public override void Write(string message)
+                        stream.Position = stream.Length;
+                        writer.Write(message);
+                    }
+                }
```
I think this could just be:
```C#
File.AppendAllText(LogFileName, message);
```

stephentoub commented on src/System.Diagnostics.TraceSource/tests/DefaultTraceListenerClassTests.cs:
```diff
@@ -27,6 +29,12 @@ public override void Write(string message)
+            public override void WriteLine(string message)
+            {
+                if (ShouldOverrideWriteLine)
+                    _writer.WriteLine(message);
```
Is this missing:
```C#
else
{
    base.WriteLine(message);
}
```
?

safern replied:
I wasn't needed, but to test the log file functionality that was added it is needed. Will add it with the other tests.

stephentoub commented on src/System.Diagnostics.TraceSource/tests/DefaultTraceListenerClassTests.cs:
```diff
@@ -45,6 +53,16 @@ public void Fail()
+            Assert.False(listener.Output.Contains("FAIL"));
+        }
+
+        [Fact]
```
Do we have any tests that validate the log file that's generated?  Seems like probably not, since it looks like that functionality was only just added.  Can we add some?

safern replied:
Yes, will add some tests to validate that!!

stephentoub commented on src/System.Diagnostics.TraceSource/tests/DefaultTraceListenerClassTests.cs:
```diff
@@ -45,6 +53,16 @@ public void Fail()
         }
 
         [Fact]
+        public void Fail_WithoutWriteLineOverride()
```
If you haven't already done so, it'd be good to look at a code coverage report for the library with your changes and see what functionality you've touched isn't covered, then add tests to cover at least those lines/branches.

safern commented:
@dotnet-bot test code coverage please.

safern commented:
Thanks for your reviews @stephentoub @tarekgh.

Even though you both approved the PR, I wanted to make sure to add missing tests for the log file functionality that was added with the updates in `DefaultTraceListener`. So I've added those tests and addressed Steve's feedback. I'm running a test code coverage so you can see it as after running it locally it showed that with the tests added everything that I've touched is covered.

safern commented:
Code coverage for the apis that I've touched can be found [here](https://ci.dot.net/job/dotnet_corefx/job/master/job/code_coverage_windows_prtest/41/Code_Coverage_Report/) -> `System.Diagnostics.TraceSource` -> `DefaultTraceListener` 

cc: @stephentoub 

Is it now good to merge?

stephentoub commented:
@dotnet-bot test code coverage please

stephentoub commented on src/System.Diagnostics.TraceSource/tests/DefaultTraceListenerClassTests.cs:
```diff
@@ -45,6 +60,46 @@ public void Fail()
+            var listener = new TestDefaultTraceListener();
+            listener.LogFileName = "LogFile.txt";
+
+            string pathToLogFile = Path.Combine(Path.GetDirectoryName(Assembly.GetEntryAssembly().Location), listener.LogFileName);
```
It'd be better not to have the log file go to the execution directory, a) because it clutters up the dev's execution environment, and b) because I believe that won't work in certain environments (cc: @danmosemsft, who just hit this on some other tests).  I assume LogFileName can be set to an absolute path?  In that case I suggest having the test class derive from FileCleanupTestBase (you can see examples of this in other test classes), and then do:
```C#
string pathToLogFile = GetTestFilePath();
listener.LogFileName = pathToLogFile;
```
You then also don't need to delete it later, as that's handled by the base class.

safern replied:
Thanks for pointing me out to that class. Will do that :)

stephentoub commented on src/System.Diagnostics.TraceSource/tests/DefaultTraceListenerClassTests.cs:
```diff
@@ -45,6 +60,46 @@ public void Fail()
+        {
+            // Exception should be handled by DefaultTraceListener.WriteLine so no need to assert.
+            var listener = new TestDefaultTraceListener();
+            listener.LogFileName = @"testfiles\LogFile.txt";
```
Nit: consider using a Guid instead of "testfiles", as the latter is a name that could potentially exist at some point, e.g.
```C#
listener.LogFileName = $"{Guid.NewGuid().ToString("N")}\\LogFile.txt";
```

safern replied:
Sounds good! Thanks!

stephentoub commented:
Left a few more comments, otherwise LGTM. Thanks.

danmosemsft commented on src/System.Diagnostics.TraceSource/src/Resources/Strings.resx:
```diff
@@ -58,6 +58,9 @@
     <value>System.Resources.ResXResourceWriter, System.Windows.Forms, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089</value>
   </resheader>
+  <data name="ExceptionOccurred" xml:space="preserve">
+    <value>An exception ocurred while writing to the log file {0}: {1}.</value>
```
Typo -- occurred not ocurred

danmosemsft replied:
Feel free to fix in a followup

safern replied:
Will fix now that I'm addressing Stephen's comments! Thanks

danmosemsft commented:
@dotnet-bot test code coverage windows debug (https://github.com/dotnet/corefx/issues/16283)

safern commented:
Thanks for your reviews.
